### DISCUSSION TITLE
tsserver performance with generated zod prisma schemas is terribly slow

### DISCUSSION THREAD

**Discussion by arnm** [omar-dulaimi/prisma-trpc-generator#69] [Q&A]
How is everyone working with these generation libraries and maintaining a decent developer experience?

I was using another zod schema generation library (https://github.com/chrishoermann/zod-prisma-types/issues/89) and check this one out but I am having the same developer experience issues with this one as well.

Am I the only one noticing these slowdowns? I dont think I have a terribly large prisma schema either with 12 models.

I tried typescript references to see if organizing the generated code that way would be of any benefit but it was not.

Any one have any ideas?

**Comment by omar-dulaimi** [comment]
Maybe try to run the generation only when you need it?
Or just don't generate everything. The less you generate, the better.
This PR https://github.com/omar-dulaimi/prisma-trpc-generator/pull/68
Added support for excluding certain operations from routers. I'm planning to make Zod Generator respect these choices as well, so that schemas will get skipped.

**Reply by arnm** [reply]
Part of my issue was with actual Zod itself which they have addressed here: https://github.com/colinhacks/zod/issues/2142. This has taken my projects tsserver diagnostics run from 200 seconds  to run to about 90 seconds so still not great. I will try to remove some of the models like you suggest

**Comment by omar-dulaimi** [comment]
Closing this with a straight answer rather than leaving it open for another two years.

**@maxime4000 was right and I was wrong.** I said it was TypeScript struggling with many files and proposed bundling everything into one or two files as the fix. That would not have helped. The cost is Zod's inference: each schema builds a large structural type, and `z.infer` on a deep Prisma input type expands into something the language server has to keep re-checking. Putting the same types in fewer files does not reduce the work, and can make it worse, since one enormous file invalidates as a unit.

I never shipped that bundling option, which in hindsight is the right outcome for the wrong reason.

**What actually helps, in order of effect:**

1. **Generate less.** Operation-level exclusion landed in #68 and is the single biggest lever. Every operation you do not generate is a whole family of input types the checker never sees. If you only ever call `findMany` and `create` on a model, say so.
2. **Do not re-generate on every save.** Run it on schema change, not as part of the dev loop.
3. **`skipLibCheck: true`**, if you do not already have it.
4. Project references did not help you, and that matches the diagnosis: they partition compilation, not the inference cost inside one graph.

Twelve models is not a large schema, and you were right that it should not have felt like that. The honest position is that this generator's default of emitting every operation for every model is the problem, and narrowing what you generate is the real fix rather than a workaround.

Worth knowing if you are still on it: `prisma-zod-generator` has since gained `useMultipleFiles`, mode presets that emit far less by default, and variant-level control. It is the more actively developed of the two.
